Привет, меня зовут .... Сегодня я проверю твой проект.
<br> Дальнейшее общение будет происходить на "ты" если это не вызывает никаких проблем.
<br> Желательно реагировать на каждый мой комментарий ('исправил', 'не понятно как исправить ошибку', ...)
<br> Пожалуйста, не удаляй комментарии ревьюера, так как они повышают качество повторного ревью.

Комментарии будут в <font color='green'>зеленой</font>, <font color='blue'>синей</font> или <font color='red'>красной</font> рамках:

<div class="alert alert-block alert-success">
<b>Успех:</b> Если все сделано отлично
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> Если можно немного улучшить
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка:</b> Если требуются исправления. Работа не может быть принята с красными комментариями.
</div>

-------------------

Будет очень хорошо, если ты будешь помечать свои действия следующим образом:
<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> ...
</div>

<div class="alert alert-block alert-warning">
<b>Изменения:</b> Были внесены следующие изменения ...
</div>

<font color='green'><b>Полезные (и просто интересные) материалы:</b></font> \
Для работы с текстами используют и другие подходы. Например, сейчас активно используются RNN (LSTM) и трансформеры (BERT и другие с улицы Сезам, например, ELMO). НО! Они не являются панацеей, не всегда они нужны, так как и TF-IDF или Word2Vec + модели из классического ML тоже могут справляться. \
BERT тяжелый, существует много его вариаций для разных задач, есть готовые модели, есть надстройки над библиотекой transformers. Если, обучать BERT на GPU (можно в Google Colab или Kaggle), то должно быть побыстрее.\
https://huggingface.co/transformers/model_doc/bert.html \
https://t.me/renat_alimbekov \
https://colah.github.io/posts/2015-08-Understanding-LSTMs/ - Про LSTM \
https://web.stanford.edu/~jurafsky/slp3/10.pdf - про энкодер-декодер модели, этеншены\
https://pytorch.org/tutorials/beginner/transformer_tutorial.html - официальный гайд
по трансформеру от создателей pytorch\
https://transformer.huggingface.co/ - поболтать с трансформером \
Библиотеки: allennlp, fairseq, transformers, tensorflow-text — множествореализованных
методов для трансформеров методов NLP \
Word2Vec https://radimrehurek.com/gensim/models/word2vec.html 

<font color='green'>Пример BERT с GPU:
```python
%%time
from tqdm import notebook
batch_size = 2 # для примера возьмем такой батч, где будет всего две строки датасета
embeddings = [] 
for i in notebook.tqdm(range(input_ids.shape[0] // batch_size)):
        batch = torch.LongTensor(input_ids[batch_size*i:batch_size*(i+1)]).cuda() # закидываем тензор на GPU
        attention_mask_batch = torch.LongTensor(attention_mask[batch_size*i:batch_size*(i+1)]).cuda()
        
        with torch.no_grad():
            model.cuda()
            batch_embeddings = model(batch, attention_mask=attention_mask_batch)
        
        embeddings.append(batch_embeddings[0][:,0,:].cpu().numpy()) # перевод обратно на проц, чтобы в нумпай кинуть
        del batch
        del attention_mask_batch
        del batch_embeddings
        
features = np.concatenate(embeddings) 
```
Можно сделать предварительную проверку на наличие GPU.\
Например, так: ```device = torch.device("cuda:0") if torch.cuda.is_available() else torch.device("cpu")```\
Тогда вместо .cuda() нужно писать .to(device)

Если понравилась работа с текстами, то можешь посмотреть очень интересный (но очень-очень сложный) курс лекций: https://github.com/yandexdataschool/nlp_course .
</font>

### <font color='orange'>Общее впечатление</font>
* Большое спасибо за проделанную работу. Видно, что приложено много усилий.
* Радует, что ноутбук хорошо структурирован. Приятно проверять такие работы.
* Работа получилась отличной, тебе удалось добиться достаточно хорошего качества. Поздравляю!
* Проект может быть зачтен, но я его отправлю назад, чтобы у тебя была возможность задать вопросы и внести правки, при желании. Однако, ты можешь просто вернуть проект в таком же виде и я его зачту.

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> Спасибо за материалы
</div>


### <font color='orange'>Общее впечатление (ревью 2)</font>
* Проект зачтен!
* Удачи в дальнейшем обучении и следующих работах!

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Общее-впечатление" data-toc-modified-id="Общее-впечатление-0.1"><span class="toc-item-num">0.1&nbsp;&nbsp;</span><font color="orange">Общее впечатление</font></a></span></li><li><span><a href="#Общее-впечатление-(ревью-2)" data-toc-modified-id="Общее-впечатление-(ревью-2)-0.2"><span class="toc-item-num">0.2&nbsp;&nbsp;</span><font color="orange">Общее впечатление (ревью 2)</font></a></span></li></ul></li><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка</a></span></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></div>

# Проект для «Викишоп»

Интернет-магазину «Викишоп» нужен инструмент, который будет искать токсичные комментарии в описанияя товаров отправлять их на модерацию. 
В распоряжении набор размеченных данных.

Необходимо:
1. Обучить модель классифицирующую комментарии на позитивные и негативные. 
2. Построить модель со значением метрики качества *F1* не меньше 0.75. 


## Подготовка

In [1]:
import pandas as pd
from catboost import CatBoostClassifier
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import f1_score, precision_recall_curve

import re
import string

<div class="alert alert-block alert-success">
<b>Успех:</b> Отлично, что все импорты собраны в первой ячейке ноутбука! Если у того, кто будет запускать твой ноутбук будут отсутствовать некоторые библиотеки, то он это увидит сразу, а не в процессе!
</div>

In [2]:
df = pd.read_csv('/datasets/toxic_comments.csv')
df.head()

text  toxic
0  Explanation\nWhy the edits made under my usern...      0
1  D'aww! He matches this background colour I'm s...      0
2  Hey man, I'm really not trying to edit war. It...      0
3  "\nMore\nI can't make any real suggestions on ...      0
4  You, sir, are my hero. Any chance you remember...      0

Столбец *text* в нём содержит текст комментария, а *toxic* — целевой признак.

In [3]:
df.toxic.value_counts(normalize=True)

0    0.898321
1    0.101679
Name: toxic, dtype: float64

Токсичных коментариев в выборке около 10%

<div class="alert alert-block alert-success">
<b>Успех:</b> Данные загружены корректно. Радует, что абаланс классов был изучен!
</div>

Необходимо очистить данные от символов, цифр, ссылок и т.п. А также привести слова в строчный шрифт.

In [4]:
def text_cleaning(text):
    text = text.lower()
    text = re.sub('\[.*?\]', '', text)
    text = re.sub("\'", '', text)
    text = re.sub('\\W',' ',text)
    text = re.sub('https?://\S+|www\.\S+', '', text)
    text = re.sub('<.*?>+', '', text)
    text = re.sub('[%s]' % re.escape(string.punctuation), '', text)
    text = re.sub('\n', '', text)
    text = re.sub('\w*\d\w*', '', text)
    
    return text

In [5]:
df['text']=df['text'].apply(text_cleaning)
df.head()

text  toxic
0  explanation why the edits made under my userna...      0
1  daww  he matches this background colour im see...      0
2  hey man  im really not trying to edit war  its...      0
3    more i cant make any real suggestions on imp...      0
4  you  sir  are my hero  any chance you remember...      0

<div class="alert alert-block alert-success">
<b>Успех:</b> Очистка была сделана верно.
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> желательно было еще провести лемматизацию.
</div>

## Обучение

Разбиваем данные на тренировочные, валидационные и тестовые.

In [6]:
X_train, X_test, y_train, y_test = train_test_split(df.drop('toxic', axis=1),
                                                    df.toxic,
                                                    test_size=0.2,
                                                    random_state=42)
X_train, X_valid, y_train, y_valid = train_test_split(X_train,
                                                      y_train,
                                                      test_size=0.1,
                                                      random_state=42)

Оценим может ли CatBoostClassifier обеспечить необходимую точность

In [7]:
cbc = CatBoostClassifier(iterations=1000,                         
                         eval_metric='F1',                        
                         text_features=['text'],
                         random_seed=42,
                         verbose=500)

cross_val_score(cbc, X_train, y_train, cv=3, scoring='f1').mean()

Learning rate set to 0.065688
0:	learn: 0.6928900	total: 802ms	remaining: 13m 21s
500:	learn: 0.7876905	total: 5m 28s	remaining: 5m 27s
999:	learn: 0.8200429	total: 11m 19s	remaining: 0us
Learning rate set to 0.065688
0:	learn: 0.7116975	total: 783ms	remaining: 13m 2s
500:	learn: 0.7911252	total: 5m 51s	remaining: 5m 50s
999:	learn: 0.8189391	total: 11m 45s	remaining: 0us
Learning rate set to 0.065688
0:	learn: 0.6888844	total: 752ms	remaining: 12m 31s
500:	learn: 0.7865735	total: 6m 10s	remaining: 6m 8s
999:	learn: 0.8162390	total: 12m 6s	remaining: 0us


0.7739772068106703

То, что надо.

Обучаем модель

In [8]:
cbc = CatBoostClassifier(iterations=5000,                         
                         eval_metric='F1',                        
                         text_features=['text'],
                         random_seed=42,
                         use_best_model=True,
                         verbose=1000)

cbc.fit(X_train, y_train, eval_set=(X_valid, y_valid))
f1_score(y_test, cbc.predict(X_test)) 

Learning rate set to 0.050786
0:	learn: 0.7074892	test: 0.7436415	best: 0.7436415 (0)	total: 902ms	remaining: 1h 15m 7s
1000:	learn: 0.8032460	test: 0.7937034	best: 0.7943615 (879)	total: 14m 35s	remaining: 58m 18s
2000:	learn: 0.8310244	test: 0.7929043	best: 0.7953700 (1578)	total: 28m 58s	remaining: 43m 25s
3000:	learn: 0.8471481	test: 0.7915980	best: 0.7953700 (1578)	total: 42m 42s	remaining: 28m 27s
4000:	learn: 0.8628601	test: 0.7909465	best: 0.7953700 (1578)	total: 56m 6s	remaining: 14m
4999:	learn: 0.8775888	test: 0.7902761	best: 0.7953700 (1578)	total: 1h 9m 58s	remaining: 0us

bestTest = 0.7953699876
bestIteration = 1578

Shrink model to first 1579 iterations.


0.7845812723861504

Повышаем точность за счет выбора наилучшего threshold

In [11]:
precision, recall, thresholds = precision_recall_curve(y_train, cbc.predict_proba(X_train)[:, 1])
result = pd.DataFrame([precision,recall,thresholds]).T
result.columns = ['precision','recall','threshold']
result['f1'] = 2*result.precision*result.recall/(result.precision+result.recall)
boundary = result[result.f1 == result.f1.max()].threshold.values

In [12]:
f1_score(y_valid, (cbc.predict_proba(X_valid)[:, 1]>boundary)*1)

0.7992063492063491

<div class="alert alert-block alert-success">
<b>Успех:</b> Разбиение было сделано верно. Хороший подход с использованием внутреннего метода коидровки!
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> Можно было еще нестроить параметры с помощью кросс-валидации.
</div>

## Вывод

Для распознования токсичных комментариев удалось построить модель CatBoostClassifier со значеним метрики F1 выше 0.75 (0.799)